src/corppa/ocr/gvision_ocr.py

```diff
@@ -1,66 +1,293 @@
+
+# Attempt to import Google Cloud Vision Python Client
+try:
+    from google.cloud import vision
+except ImportError:
+    print(
+        "Error: Python environment does not contain google-cloud-vision "
+        "package. Switch environments or install package and try again."
+    )
+    sys.exit(1)
 
-from google.cloud import vision
+# Workaround (hopefully temporary) to surpress some logging printed to stderr
+os.environ["GRPC_VERBOSITY"] = "NONE"
 
 
-IMAGE_DIR = "images"
+def image_relpath_generator(image_dir, ext_set, follow_symlinks=True):
+    """
+    This generator method finds all images in image_dir with file extensions
+    in ext_set. For each of these images, the method yields the relative path
+    with respect to image_dir.
 
+    For example, if image_dir = "a/b/c/images" and there are image files at the
+    following paths: "a/b/c/images/alpha.jpg", "a/b/c/images/d/beta.jpg"
+    The generate will produce these two items: "alpha.jpg" and "d/beta.jpg"
+    """
+    # Using pathlib.walk over blob because (1) it allows us to find files with
+    # multiple extensions in a single walk of the directory and (2) lets us
+    # leverage additional functionality of pathlib.
+    for dirpath, dirs, files in image_dir.walk(follow_symlinks=follow_symlinks):
+        # Check the files in walked directory
+        for file in files:
+            ext = os.path.splitext(file)[1]
+            if ext:
+                filepath = dirpath.joinpath(file)
+                yield filepath.relative_to(image_dir)
+        # For future walking, remove hidden directories
+        dirs[:] = [d for d in dirs if d[0] != "."]
 
-def get_image_ocr():
+
+def ocr_images(in_dir, out_dir, ext_list, ocr_limit=0, show_progress=True):
+    """
+    OCR images in in_dir with extension ext_list to out_dir. If ocr_limit > 0,
+    stop after OCRing ocr_limit images.
+
+    Returns a map structure reporting the number of images OCR'd and skipped.
+    """
     # Instantiate google vision client
     client = vision.ImageAnnotatorClient()
 
-    for imagefile in glob.iglob("%s/*/*.jpg" % IMAGE_DIR):
-        basepath = os.path.splitext(imagefile)[0]
-        textfile = "%s.txt" % basepath
-        jsonfile = "%s.json" % basepath
-
-        # if text file already exists, skip
-        if os.path.isfile(textfile):
-            continue
+    # Setup up progress bar if progress will be shown
+    if show_progress:
+        desc = "OCRing images"
+        maxinterval = 1
+        if ocr_limit:
+            progress_bar = tqdm(desc=desc, total=ocr_limit, maxinterval=maxinterval)
+        else:
+            bar_format = "{desc}: {n:,} images OCR'd | elapsed: {elapsed}, {rate_fmt}"
+            progress_bar = tqdm(
+                desc=desc, bar_format=bar_format, maxinterval=maxinterval
+            )
 
-        # if text file does not exist, request ocr
+    ocr_count = 0
+    skip_count = 0
+    for image_relpath in image_relpath_generator(in_dir, set(ext_list)):
```

Comment:
did you consider wrapping tqdm around the image generator? might simplify the code slightly

Reply:
Honestly, I don't quite remember; I think this might be a holdover from when I there wasn't a standalone generator.